Add IHP2 Analysis Scripts to Repo

AndrewF70 opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

Upload generalized version of the analysis script used to generate graphs for IHP2 Validation Testing. Also included are scripts for installing dependencies and decompressing bag files.

## Related GitHub Issue

<!--- This project only accepts pull requests related to open GitHub issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key

<!-- e.g. CAR-123 -->

## Motivation and Context

The reason for uploading this code is to release a generalized version of the analysis script that allows the user to input IHP2 testing bag files and create graphs based on the data. The analysis script was originally developed by Andy Lam (VOLPE), and modified by Andrew Fortier (Leidos), Ankur Tyagi (Leidos), and Andy Gaines (Leidos).

## How Has This Been Tested?

The "IHP2_Analysis_Script.ipynb" script has been tested on bag files from IHP2 testing to confirm functionality, and to confirm that the graphs produced match graphs in the draft IHP2 Validation Testing report. The "install_dependencies.sh" script has been tested, and the "bag_decompress.sh" script has been tested on bag files from IHP2 testing. Andy Gaines has peer reviewed and tested the different scripts, and updates were made based on feedback.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

SaikrishnaBairamoni committed:
add Ihp2 scripts

AndrewF70 committed:
Update README.md

Add attributions

AndrewF70 committed:
Update README.md

Update formatting

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/README.md:
```diff
@@ -0,0 +1,23 @@
+The data analysis script in this repository is based on testing completed for the Integrated Highway Prototype II (IHP2) system, and that is documented in the Integrated Highway Prototype II System Validation Report.
```
If the report has been published, can we include a link to it? Alternatively if it is only available internally can we include a link to that and state that it is only available to internal Leidos employees? This would also be helpful information to include in the PR description.

AndrewF70 replied:
The report is not yet published - I'm not certain if there is a current storage location accessible via internal link for the draft report. I did add a comment to the README to clarify the report is not yet published, though please let me know anything further needed for this.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/README.md:
```diff
@@ -0,0 +1,23 @@
+The data analysis script in this repository is based on testing completed for the Integrated Highway Prototype II (IHP2) system, and that is documented in the Integrated Highway Prototype II System Validation Report.
```
Nit: Can we change from "in this repository" to "in this directory"?

AndrewF70 replied:
Completed

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/README.md:
```diff
@@ -0,0 +1,23 @@
+The data analysis script in this repository is based on testing completed for the Integrated Highway Prototype II (IHP2) system, and that is documented in the Integrated Highway Prototype II System Validation Report.
```
Since this repository has many data analysis scripts corresponding to different software releases, it can be helpful to state the approximate date that tests were conducted. Can we work into this text that this script is intended for ROS1 rosbags collected from this testing, which occurred in Summer 2022?

AndrewF70 replied:
Added clarifying sentence for this point.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/README.md:
```diff
@@ -0,0 +1,23 @@
+The data analysis script in this repository is based on testing completed for the Integrated Highway Prototype II (IHP2) system, and that is documented in the Integrated Highway Prototype II System Validation Report.
```
Can we specifically mention that the ihp2_data_analysis_script.ipynb contains the data analysis script, and that it is setup as a Jupyter Notebook?

AndrewF70 replied:
Added clarifying sentence to cover this.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/README.md:
```diff
@@ -0,0 +1,23 @@
+
+sudo bash install_dependencies.sh
+
+Step 2: If needed, decompress the bag files you will be using. To do so, run the script bag_decompress.sh in terminal by navigating to the directory with the script file and running the command below (make sure the bag files are in the same directory as the script):
```
I could also be worth mentioning that this script will save all original rosbag files in <path_to_script>/origbag

AndrewF70 replied:
Added a note to mention this point.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+{
```
Nit: I believe Jupter Notebooks are typically lowercase, consider renaming to ihp2_analysis_scrip.ipynb

AndrewF70 replied:
File renamed.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#Use this section for 3 vehicle test run\n",
```
[PEP 8](https://peps.python.org/pep-0008) is a style guide for Python and can be a good reference. This is just a nit, but consider adding a space between '#' and your comments to improve readability (reference: [PEP 8 - Inline Comments](https://peps.python.org/pep-0008/#inline-comments)

```
#Incorrect
# Correct
```

AndrewF70 replied:
Comment spacing updated

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "test_number = \"\" #Input test number in format \"T##\"\n",
+    "run_number = \"\" #Input run number in format \"R#\"\n",
+    "\n",
+    "if lead_v1_file_read != \"\":\n",
```
Consider adding errors logs or print statements in case a file is not found (due to a typo in the provide file names or the file does not exist)

AndrewF70 replied:
Error statements added for exception when file is not found

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "#3 vehicle test run\n",
+    "if three_vehicle_bool == True :\n",
+    "\n",
+    "      # delay time until start carma engaged\n",
```
This cell seems to just be extracting the timestamp that each vehicle's CARMA Platform system was engaged (though it might be used to delay some time value later on in the notebook), could we revise this comment?

AndrewF70 replied:
Comment revised

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "if three_vehicle_bool == True :\n",
+    "\n",
+    "      # delay time until start carma engaged\n",
+    "      lead_v1_engaged_flag = 0\n",
```
Nit: "flag" is commonly used to indicate a boolean variable, whereas this variable is being used to store a timestamp. Consider renaming these variables to better indicate that a timestamp is stored within them.

AndrewF70 replied:
Variable renamed

JonSmet replied:
It looks like there are still some *_flag variables in the Jupyter Notebook that could be changed to *_time, could you take a look?

AndrewF70 replied:
Renamed the remaining flag variables to include *_time.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "    #call for follower_v3 vehicle speed\n",
+    "    follower_v3_VehSpeed_dict = {}\n",
+    "    for topic, msg, t in follower_v3_bag.read_messages(topics = [\"/hardware_interface/vehicle/twist\"], start_time=follower_v3_engaged_flag):\n",
+    "        follower_v3_VehSpeed_dict[t.to_sec()] = msg.twist.linear.x  \n",
```
It might be worth adding a comment to these stating that <msg>.twist.linear.x is the vehicle's longitudinal velocity in m/s.

AndrewF70 replied:
Comments added to clarify

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "    follower_v2_AdvisorySpeed_df = pd.DataFrame.from_dict(follower_v2_AdvisorySpeed_dict, orient='index').rename(columns={0:'follower_v2_AdvisorySpeed'}).rename_axis(\"Time\").reset_index()\n",
+    "    follower_v2_AdvisorySpeed_df['DateTime'] = pd.to_datetime(follower_v2_AdvisorySpeed_df['Time'],unit='s')\n",
+    "\n",
+    "    follower_v2_StartPlatooning = pd.to_datetime(follower_v2_platoon_flag.to_sec(),unit='s')\n",
```
It could be helpful to distinguish that these are timestamps

AndrewF70 replied:
Variable renamed

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+   "outputs": [],
+   "source": [
+    "#3 vehicle test run\n",
+    "if three_vehicle_bool == True :\n",
```
Nit: Referencing [PEP 8 - Programming Recommendations](https://peps.python.org/pep-0008/#programming-recommendations), consider not comparing booleans to True or False and instead rewriting as just `if three_vehicle_bool:` (applies to other instances of this in the script as well)

AndrewF70 replied:
Updated format to match PEP 8 guidelines

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "    lead_v1_VehSpeed_df = pd.DataFrame.from_dict(lead_v1_VehSpeed_dict, orient = 'index').rename(columns = {0:'lead_v1_VehSpeed'}).rename_axis(\"Time\").reset_index()\n",
+    "    lead_v1_VehSpeed_df['DateTime'] = pd.to_datetime(lead_v1_VehSpeed_df['Time'],unit='s')\n",
+    "\n",
+    "    #CALL FOR ACTUAL GAP\n",
```
Nit: Consider changing to lowercase for consistency

AndrewF70 replied:
Updated to lowercase

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "        follower_v2_RequestPlatoon_dict[t.to_sec()] = msg.m_header.timestamp  \n",
+    "    follower_v2_RequestPlatoon_df = pd.DataFrame.from_dict(follower_v2_RequestPlatoon_dict, orient = 'index').rename(columns = {0:'follower_v2_RequestTime'}).rename_axis(\"Time\").reset_index()\n",
+    "    follower_v2_RequestPlatoon_df['DateTime'] = pd.to_datetime(follower_v2_RequestPlatoon_df['Time'],unit='s')\n",
+    "    #first instance of an “incoming mobility request” \n",
```
I might be misinterpreting this, but I think this should be "outgoing mobility request"?

AndrewF70 replied:
I believe this is correct based on the topic name - comments updated (commit pending)

AndrewF70 replied:
Comments updated

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "    follower_v3_ActualGap_df = pd.DataFrame.from_dict(follower_v3_ActualGap_dict, orient='index').rename(columns={0:'follower_v3_ActualGap'}).rename_axis(\"Time\").reset_index()\n",
+    "    follower_v3_ActualGap_df['DateTime'] = pd.to_datetime(follower_v3_ActualGap_df['Time'],unit='s')\n",
+    "\n",
+    "    #Platooning\n",
```
Consider renaming to `# Obtain timestamp at which each vehicle first requests to join a platoon` (or similar)

AndrewF70 replied:
Comment updated to "Obtain timestamp at which each vehicle first requests to join platoon"

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "#3 vehicle test run\n",
+    "if three_vehicle_bool == True :\n",
+    "\n",
+    "    #call for follower_v3 vehicle speed\n",
```
There are many lines of code that exist between this cell and the previous cell, consider combining them into functions that we can place at the top of the Jupyter Notebook so that we can call them as needed to minimize code duplication

AndrewF70 replied:
Have created a draft version of script that incorporates functions, though requires implementing for all applicable lines of code and testing to verify functionality/no issues are created. As discussed, will be planned to implement this change in a future improvement.

AndrewF70 replied:
Added functions for repeated lines of code.

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/IHP2_Analysis_Script.ipynb:
```diff
@@ -0,0 +1,534 @@
+    "      # delay time until start carma engaged\n",
+    "      lead_v1_engaged_flag = 0\n",
+    "      for topic, msg, t in lead_v1_bag.read_messages(topics = [\"/guidance/state\"]):\n",
+    "            if msg.state == 4:\n",
```
Consider adding a comment stating that '4' indicates that the CARMA Platform system is engaged

AndrewF70 replied:
Comments added to clarify

AndrewF70 committed:
Update README.md

Updates to README file based on PR #115 review feedback from Jon Smet.  Additional background added about IHP2 Validation Testing and intended rosbag files for script. Attribution updated.

AndrewF70 committed:
Rename IHP2_Analysis_Script.ipynb to ihp2_analysis_script.ipynb

Change file name to lower case notation used for Jupyter Notebook files.

AndrewF70 committed:
Update analysis script with variable names, etc.

AndrewF70 committed:
Correction to comment spacing

JonSmet commented on src/rosbag_analysis_tools/scripts/IHP2_analysis_scripts/ihp2_analysis_script.ipynb:
```diff
@@ -0,0 +1,565 @@
+    "    follower_v2_RequestPlatoon_df = pd.DataFrame.from_dict(follower_v2_RequestPlatoon_dict, orient = 'index').rename(columns = {0:'follower_v2_RequestTime'}).rename_axis(\"Time\").reset_index()\n",
+    "    follower_v2_RequestPlatoon_df['DateTime'] = pd.to_datetime(follower_v2_RequestPlatoon_df['Time'],unit='s')\n",
+    "    # first instance of an “outgoing mobility request” \n",
+    "    follower_v2_RequestPlatooning = (follower_v2_RequestPlatoon_df['DateTime'].iloc[0])\n",
```
For these `*_RequestPlatooning` variables, could we rename to `*_RequestPlatooning_time` to indicate that they are timestamps? Please take a look at other variables in the script as well in case there are others worth updating

AndrewF70 replied:
Added *_time to all additional timestamp variables.

AndrewF70 committed:
Renaming of flag variables, and updating all timestamp variables to include _time in name. Addition of functions to script to replace repeated lines of code.

JonSmet reviewed (approved):
Looks good! Thanks for addressing all of these comments.
